cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll

```diff
@@ -0,0 +1,590 @@
+ */
+private class NarrowingCastInstruction extends ConvertInstruction {
+  NarrowingCastInstruction() {
+    not this instanceof SafeCastInstruction and
+    typeBound(getResultType(), _, _)
+  }
+  /** Gets the lower bound of the resulting type. */
+  int getLowerBound() { typeBound(getResultType(), result, _) }
+  /** Gets the upper bound of the resulting type. */
+  int getUpperBound() { typeBound(getResultType(), _, result) }
+}
+
+/**
+ * Holds if `op + delta` is a valid bound for `i`.
+ * - `upper = true`  : `i <= op + delta`
+ * - `upper = false` : `i >= op + delta`
+ */
+private predicate boundFlowStep(Instruction i, Operand op, int delta, boolean upper) {
+  valueFlowStep(i, op, delta) and
+  (upper = true or upper = false)
+  or
+  i.(SafeCastInstruction).getAnOperand() = op and
+  delta = 0 and
+  (upper = true or upper = false)
+  or
+  exists(Operand x |
+    i.(AddInstruction).getAnOperand() = op and
+    i.(AddInstruction).getAnOperand() = x and
+    op != x
+    |
+    not exists(getValue(getConstantValue(op.getInstruction()))) and
+    not exists(getValue(getConstantValue(x.getInstruction()))) and
+    if(strictlyPositive(x))
+    then (
+      upper = false and delta = 1
+    ) else
+      if positive(x)
+      then (
+        upper = false and delta = 0
+      ) else
+        if strictlyNegative(x)
+        then (
+          upper = true and delta = -1
+        ) else if negative(x) then (upper = true and delta = 0) else none()
+  )
+  or
+  exists(Operand x |
+    exists(SubInstruction sub |
+      i = sub and
+      sub.getAnOperand().(LeftOperand) = op and
+      sub.getAnOperand().(RightOperand) = x
+    )
+  |
+    // `x` with constant value is covered by valueFlowStep
+    not exists(getValue(getConstantValue(x.getInstruction()))) and
+    if strictlyPositive(x)
+    then (
+      upper = true and delta = -1
+    ) else
+      if positive(x)
+      then (
+        upper = true and delta = 0
+      ) else
+        if strictlyNegative(x)
+        then (
+          upper = false and delta = 1
+        ) else if negative(x) then (upper = false and delta = 0) else none()
+  )
+  or
+  i.(RemInstruction).getAnOperand().(RightOperand) = op and positive(op) and delta = -1 and upper = true
+  or
+  i.(RemInstruction).getAnOperand().(LeftOperand) = op and positive(op) and delta = 0 and upper = true
+  or
+  i.(BitAndInstruction).getAnOperand() = op and positive(op) and delta = 0 and upper = true
+  or
+  i.(BitOrInstruction).getAnOperand() = op and positiveInstruction(i) and delta = 0 and upper = false
+  // TODO: min, max, rand
+}
+
+private predicate boundFlowStepMul(Instruction i1, Operand op, int factor) {
+  exists(Instruction c, int k | k = getValue(getConstantValue(c)) and k > 0 |
+    i1.(MulInstruction).hasOperands(op, c.getAUse()) and factor = k
+    or
+    exists(ShiftLeftInstruction i |
+      i = i1 and i.getAnOperand().(LeftOperand) = op and i.getAnOperand().(RightOperand) = c.getAUse() and factor = 2.pow(k)
```

Comment:
It seems like maybe adding `getLeftOperand()` on `Instruction` as a short-hand for `.getAnOperand().(LeftOperand)` would be nice to have (resp. `getRightOperand()`).

Reply:
Or maybe I'm missing something here, because I immediately notice the use of `i.getRightOperand()` below.  Is there a difference between `i.getRightOperand()` and `i.getAnOperand().(RightOperand)`, seeing as the latter is used quite a bit?

Reply:
Yes; originally, there wasn't an `Operand` IPA type, and so `getLeftOperand` and company get an `Instruction`. `getAnOperand().getLeftOperand()` gets the `Operand` object. I think we need to do some cleanup and renaming on that API, but probably not in this PR.